daffodil-runtime2/src/main/resources/org/apache/daffodil/runtime2/c/libruntime/parsers.c

```diff
@@ -18,140 +18,508 @@
+        }
+
+        // Fill the fractional byte
+        size_t num_bits_fill = BYTE_WIDTH - pstate->unreadLen;
+        pstate->bitPos0b += num_bits_fill;
+        pstate->unreadBits <<= num_bits_fill;
+        pstate->unreadBits |= HIGH_BITS(whole_byte, num_bits_fill);
+        pstate->unreadLen += num_bits_fill;
+        whole_byte <<= num_bits_fill;
+
+        // Copy the fractional byte to `bytes`
+        size_t num_bits_read = BYTE_WIDTH;
+        if (num_bits_read > num_bits) num_bits_read = num_bits;
+        num_bits -= num_bits_read;
+        bytes[ix_bytes++] = pstate->unreadBits & HIGH_MASK(num_bits_read);
+        pstate->unreadLen -= num_bits_read;
+
+        // Copy any remaining unused bits from the whole byte to the
+        // fractional byte
+        size_t num_bits_unused = BYTE_WIDTH - num_bits_fill;
+        if (num_bits_unused)
+        {
+            pstate->bitPos0b += num_bits_unused;
+            pstate->unreadBits <<= num_bits_unused;
+            pstate->unreadBits |= HIGH_BITS(whole_byte, num_bits_unused);
+            pstate->unreadLen += num_bits_unused;
+        }
+    }
+
+    // Copy the fractional byte to `bytes` one last time
+    if (num_bits)
+    {
+        assert(num_bits <= pstate->unreadLen);
+
+        // Shift out any unused bits in the fractional byte
+        size_t num_bits_unused = BYTE_WIDTH - pstate->unreadLen;
+        pstate->unreadBits <<= num_bits_unused;
+        bytes[ix_bytes] = pstate->unreadBits & HIGH_MASK(num_bits);
+        pstate->unreadLen -= num_bits;
+
+        // Shift back any unread bits still left in fractional byte
+        pstate->unreadBits >>= num_bits_unused;
+    }
+}
+
+// Helper method to read booleans using fractional byte shifts
+// depending on data endianness (note not tested on big-endian
+// architecture; might work only on low-endian architecture)
+
+static void
+parse_endian_bool(bool big_endian_data, bool *number, size_t num_bits, int64_t true_rep, uint32_t false_rep,
+                  PState *pstate)
+{
+    // Parse all booleans in this helper function
+    union
+    {
+        uint8_t  bytes[sizeof(uint32_t)];
+        uint32_t integer;
+    } buffer;
+
+    // Read data bits
+    buffer.integer = 0;
+    read_bits(buffer.bytes, num_bits, pstate);
+
+    // Shift data bits differently on endianness
+    if (big_endian_data)
+    {
+        // Convert data endianness to host endianness
+        buffer.integer = be32toh(buffer.integer);
+
+        // Need only num_bits so remove any padding bits
+        size_t shift = sizeof(uint32_t) * BYTE_WIDTH - num_bits;
+        buffer.integer >>= shift;
+    }
+    else
+    {
+        // Convert data endianness to host endianness
+        buffer.integer = le32toh(buffer.integer);
+
+        // Need only num_bits so remove any padding bits
+        size_t msb_ix = (num_bits - 1) / BYTE_WIDTH;
+        size_t shift = (BYTE_WIDTH - num_bits % BYTE_WIDTH) % BYTE_WIDTH;
+        buffer.bytes[msb_ix] >>= shift;
+    }
```

Comment:
Can/should this call `parse_endian_uint32` and then compare the result to true/false rep? That should minimize duplicate code, right?